[Core] Support setting private workspace and adding users to it

DanielZhangQD opened the pull request:
<!-- Describe the changes in this PR -->

Support setting a private workspace and adding users to it.
Configurations:
```
 <workspace>:
    private: true
    allowed_users:
    - user1
```
- The public workspaces can be accessed by all users
- The private workspace can be accessed by users with the `User` role in the `allowed_users` and users with the `Admin` role.

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
    See separate comment below. 
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local)
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

<!-- CI commands (/-prefixed) can only be triggered by repo members -->

Michaelvll commented:
> The role update and allowed users update may take some time(60s at most) to take effect as we introduce a thread to load policy every 60s.

Since we have moved the workspace from the critical path, should we just reload everytime or reload for every request, instead of using the watch?

role update should be instant in dashboard, otherwise it could be a bad user experience.

DanielZhangQD commented on sky/utils/common_utils.py:
```diff
@@ -1011,3 +1011,31 @@ def _get_cgroup_memory_limit() -> Optional[int]:
     return os.path.isfile('/sys/fs/cgroup/cgroup.controllers')
+
+
+def get_workspace_users(workspace_config: Dict[str, Any],
```
This is required when we rebuild the workspace policy in rbac.py, so move this to common_utils.

DanielZhangQD replied:
Moved to `sky/workspaces/utils.py`

DanielZhangQD commented on sky/utils/common_utils.py:
```diff
@@ -1011,3 +1011,31 @@ def _get_cgroup_memory_limit() -> Optional[int]:
+
+
+def get_workspace_users(workspace_config: Dict[str, Any],
+                        all_users: List[models.User]) -> List[str]:
```
Yes, doing this because it cannot import global_user_state here due to circle import, do you have any better suggestions?

DanielZhangQD replied:
Seems moving it to a separate file works, let me try.

DanielZhangQD commented on test_workspace_permissions.py:
```diff
@@ -36,19 +36,16 @@ def test_public_workspace_config(self):
         """Test that public workspace config returns wildcard."""
         public_config = {'private': False}
         users = common_utils.get_workspace_users(public_config, self.all_users)
-        self.assertEqual(users, ['*'], 
```
ok, I will try to optimize the file and leave the useful cases only.

DanielZhangQD commented on sky/users/rbac.py:
```diff
@@ -89,32 +90,35 @@ def get_role_permissions(
 
 def get_workspace_policy_permissions(
-        workspace_name: Optional[str] = None) -> Dict[str, List[str]]:
+        workspace_names: Optional[str] = None) -> Dict[str, List[str]]:
```
Yes, it's not used for now, I thought you wanted to make this function more common, let me remove it then.

DanielZhangQD replied:
Removed

DanielZhangQD commented:
> > The role update and allowed users update may take some time(60s at most) to take effect as we introduce a thread to load policy every 60s.
> 
> Since we have moved the workspace from the critical path, should we just reload everytime or reload for every request, instead of using the watch?
> 
> role update should be instant in dashboard, otherwise it could be a bad user experience.

Updated to load policy each time.

DanielZhangQD commented on sky/workspaces/core.py:
```diff
@@ -520,48 +527,20 @@ def workspaces_for_user(user: models.User) -> List[str]:
     return user_workspaces
-
-
-def _get_workspace_users(workspace_config: Dict[str, Any]) -> List[str]:
```
Moved to sky/workspaces/utils.py

DanielZhangQD commented:
/smoke-test --kubernetes

DanielZhangQD commented:
/smoke-test --kubernetes

DanielZhangQD commented:
/smoke-test --gcp

DanielZhangQD commented:
Smoke test result:
- `test_long_setup_run_script` and `test_cancel_pytorch` fail on both this branch and master, traced in https://github.com/skypilot-org/skypilot/issues/5908.
- `test_multi_tenant_managed_jobs` failed, it seems related to the `user` name of the job controller cluster.
For master:
```

Clusters
--
  | NAME                          USER   INFRA                RESOURCES                              STATUS  AUTOSTOP  LAUNCHED
  | sky-jobs-controller-478b91bf  user1  GCP (us-central1-a)  1x(cpus=2, mem=8, n2-standard-2, ...)  UP      10m       5 mins ago

```
For this branch:
```

Clusters
--
  | NAME                          USER       INFRA                RESOURCES                              STATUS  AUTOSTOP  LAUNCHED
  | sky-jobs-controller-d737f77e  buildkite  GCP (us-central1-a)  1x(cpus=2, mem=8, n2-standard-2, ...)  UP      10m       6 mins ago


```

DanielZhangQD commented:
/smoke-test --kubernetes -k test_multi_tenant_managed_jobs

Michaelvll commented:
/quicktest-core

Michaelvll commented:
/smoke-test

Michaelvll commented:
/smoke-test --kubernetes

Michaelvll commented:
all smoke tests passed except for `test_long_setup_run_script` (should be fixed by #5864)

Michaelvll reviewed (changes requested):
There is still an issue with what an admin should specify for a workspace, user_hash vs user_name

Michaelvll commented:
> There is still an issue with what an admin should specify for a workspace, user_hash vs user_name

This should be resolved

Michaelvll commented on sky/users/server.py:
```diff
@@ -55,10 +51,10 @@ async def user_update(user_update_body: payloads.UserUpdateBody) -> None:
                                     detail=f'Invalid role: {role}')
-    user_info = global_user_state.get_user(user_id)
-    if not user_info.name:
+    user_info, user_exists = global_user_state.get_user_with_existence(user_id)
```
I removed the new interface. See the latest commit

DanielZhangQD replied:
OK

Michaelvll commented:
/quicktest-core

Michaelvll commented:
There seems to be an issue where some executor has been set with an active workspace when accessed through dashboard
